pkg/controller/job/job_controller.go

```diff
@@ -669,140 +670,171 @@ func newCondition(conditionType batch.JobConditionType, reason, message string)
 						errCh <- err
 					}
-				}()
+				}(i)
 			}
 			wait.Wait()
-			// any skipped pods that we never attempted to start shouldn't be expected.
-			skippedPods := diff - batchSize
-			if errorCount < len(errCh) && skippedPods > 0 {
-				glog.V(2).Infof("Slow-start failure. Skipping creation of %d pods, decrementing expectations for job %q/%q", skippedPods, job.Namespace, job.Name)
-				active -= skippedPods
-				for i := int32(0); i < skippedPods; i++ {
-					// Decrement the expected number of creates because the informer won't observe this pod
-					jm.expectations.CreationObserved(jobKey)
+
+		} else if active < parallelism {
+			wantActive := int32(0)
+			if job.Spec.Completions == nil {
+				// Job does not specify a number of completions.  Therefore, number active
+				// should be equal to parallelism, unless the job has seen at least
+				// once success, in which leave whatever is running, running.
+				if succeeded > 0 {
+					wantActive = active
+				} else {
+					wantActive = parallelism
 				}
-				// The skipped pods will be retried later. The next controller resync will
-				// retry the slow start process.
-				break
+			} else {
+				// Job specifies a specific number of completions.  Therefore, number
+				// active should not ever exceed number of remaining completions.
+				wantActive = *job.Spec.Completions - succeeded
+				if wantActive > parallelism {
+					wantActive = parallelism
+				}
+			}
+			diff := wantActive - active
+			if diff < 0 {
+				utilruntime.HandleError(fmt.Errorf("More active than wanted: job %q, want %d, have %d", jobKey, wantActive, active))
+				diff = 0
+			}
+			jm.expectations.ExpectCreations(jobKey, int(diff))
+			errCh = make(chan error, diff)
+			glog.V(4).Infof("Too few pods running job %q, need %d, creating %d", jobKey, wantActive, diff)
+
+			allocation := sync.Mutex{}
+			availableCompletionsIndexes := make([]int32, 0)
+			if completions > 0 {
+				// find available completions indexes in now
+				availableCompletionsIndexes = getAvailableCompletionsIndexes(activePods, succeededPods, completions)
+			}
+
+			active += diff
+			wait := sync.WaitGroup{}
+
+			// Batch the pod creates. Batch sizes start at SlowStartInitialBatchSize
+			// and double with each successful iteration in a kind of "slow start".
+			// This handles attempts to start large numbers of pods that would
+			// likely all fail with the same error. For example a project with a
+			// low quota that attempts to create a large number of pods will be
+			// prevented from spamming the API service with the pod create requests
+			// after one of its pods fails.  Conveniently, this also prevents the
+			// event spam that those failures would generate.
+			for batchSize := int32(integer.IntMin(int(diff), controller.SlowStartInitialBatchSize)); diff > 0; batchSize = integer.Int32Min(2*batchSize, diff) {
+				errorCount := len(errCh)
+				wait.Add(int(batchSize))
+				for i := int32(0); i < batchSize; i++ {
+					go func() {
```

Comment:
would it be hard to not use an anonymous function here?
my assumption would be "yes", but just asking as this is preferred.

Reply:
I don't quite understand what you mean, can you explain it to me?

Reply:
`go func() { /* some code */ }` runs an anonymous function in a go routine, while `go someFunc();` would run a pre-defined function called `someFunc()` in a go routine.

if a function to run in a go routine is big, it is preferred to not have it as anonymous, yet this poses the complication of passing arguments to it, and returning values from it.

ultimately, i would leave it to the maintainers for the final call.

Reply:
I think I can extract the method I generated for podTemplateSpec as a separate method. Others don't change. What do you think?

Reply:
but this case need lock and change availableCompletionsIndexes，Leaving the main method may cause the extracted method to contain side effects;
Do you have any good suggestions? @neolit123